README
Welcome to the fbled README

This project aims at providing an open-source daemon to update the front-panel
LEDs of the Watchguard Firebox II and Firebox III. These boxes are basically PC-
based appliances based on the x86 architecture. They feature 3 10/100 Ethernet
ports and a programmable LED panel.

The home of this code is
 
                              https://github.com/fmertz/fbled

Watchguard and Firebox are used here for identification purposes, and are no
doubt the property on their owner. This project has no connection whatsoever
with the manufacturer of the appliances. Other trademarks are referred to for 
identification as well.

Of direct interest for this project, these boxes feature a front panel with a number
of programmable LEDs. The LEDs are grouped in 3 stacks and a triangle. There is
a stack for specific statuses (SYS A, SYS B, ARMED, DISARMED and POWER), a stack
of 8 LEDs for LOAD and another stack of 8 LEDs for Traffic. The Triangle (meant to
have each vertex match an Ethernet port) had LEDs at each vertex, plus 2 LEDs in
each direction from all vertices.

These white LEDs can only be controlled as on or off, but can be controlled
individually. They are already behind colored masks, so they appear colored
yellow, green or red already.

The idea of the fbled project is to be able to walk up to one of these boxes and
have a sense of what the machine is up to, generally

	OFF is not applicable
	ON is normal
	BLINKING means needs attention
    BLINKING FAST means needs immediate attention

The code is logically separated into a Driver and a Client. The Driver is
responsible for updating the LEDs based on input parameters, and deals with
low-level I/O ports. The driver might be reimplemented as a real Linux driver/
module at a later point. The Client is responsible for gathering the live values
from the running system, parse them, normalize them, then pass them to the
Driver.

From a development perspective, a choice was made early on to make it
available for others to use, study and hopefully contribute. It is clear that
fbled is meant for the x86 architecture only, but it is less clear what operating
system it could run under. Therefore, in the interest of portability across OS,
this project uses the Autotools, and is basically centered around the GNU C
Library. The run-time values are gathered with portable code, but invariably
depend on the specifics of where those values come from. Following the
"everything is a file" idiom, this code follows the usual user-space interface
to the system by reading "files" in the /proc or /sys areas. This is an area
where portability issues are anticipated to crop up.

fbled is licensed under GPLv3, and is offered to others in the spirit of giving
a little something back to the free software movement.

USAGE:
fbled just runs. As any user-space process in need of control of individual I/O
port, it needs to be privileged, i.e. run as "root". This can be accomplished in
a number of ways:
- It can be started as a part of the boot process by init
- It can be started directly by the root user
- It can be stored on the file system with setuid-root, i.e. be owned by root,
  and have execution bits set to setuid. This way, it can be run by any user.
- If can be started with sudo, provided sudo is setup for the current user

OPERATING SYSTEM
Altogether, the appliance boots with a BIOS. The BIOS then boots off of IDE
storage. There is no special boot locking, so anything that can be booted
could run. There are some online guides, but basically, the jumpers can be
setup so the internal flash drive is set as slave, and a replacement, larger
capacity primary drive is setup with an image of choice. CompactFlash is easy
and cheap to setup. Real hard drives could work, too.

A quick search reveals that folks are booting a number of distributions, either
Linux-based, or BSD-based:

Linux, with the GNU C Library:
	Vyatta

Linux, with the uClibc C Library
	Open-WRT
	DD-WRT

BSD:
	pfSense
	m0n0wall

Folks are encouraged to use the wiki to share their experience and help others

                                      https://github.com/fmertz/fbled/wiki



NEWS
This NEWS file is not currently updated. The commit log on github is a better
place for the project activity.

                              https://github.com/fmertz/fbled


fbled.c
///           // fbled.c
//
// Copyright (C) 2011 - F Mertz
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

//  Wed March 16
//  Copyright  2011  F Mertz
//  <fireboxled@gmail.com>
//
// This project aims at providing a simple deamon to update the front LEDs on a Watchguard Firebox II/III. This code is
// logically separated into a Driver and a Client.
//
// The Driver is responsible for updating the LEDs based on input parameters, and deals with low-level I/O ports.
// This driver might be reimplemented as a real Linux driver/module at a later point.
//
// The Client is responsible for gathering the live values from the running system, parse them, normalize them, then pass them
// to the Driver.
//=============================================================================
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/io.h>
#include <sys/time.h>
#include <errno.h>
#include <time.h>
#include <signal.h>
#include <features.h>

//#define LED_DEBUG
#include "fbled.h"
#include "config.h"

//This array holds all LED worker functions
//	Base wait time is defined in WRK_WAIT
//	All workers are running once every uSkipCount times
//
static tExecTable tLEDProcTable[] = {
	{ .pfConstr=	NULL,	.pfCode=	DoLoad,	.pfDestr=	NULL,	.uSkipCount=	2,	.uRunningCount=	0},
	{						NULL,					DoTraffic,					NULL,							2,								1},
	{						NULL,					DoBlink,					NULL,							3,								0}
};

//This is used to control the main loop of workers
// The worker LED threads will read it, the signal handler can change it
// Turn-off any optimization by the compiler by declaring it volatile
//
static unsigned volatile uKeepGoing=1;

//This is used for stck style
static unsigned uStackStyle=STACK_BAR;

//DRIVER: Initialize function
//
static int DrvInit(unsigned char uMode)
{
	int 				iRetVal;
	unsigned	u;
	char			cEthNum;
	FILE				*tEthFile=NULL;
	char			acMACAddress[256];
	char			acFileName[256];

	//Try and find out if this code is running on a Firebox
	//Look at the MAC addresses of eth0-4 and see if at least 1 is part of the Watchguard range
	for (cEthNum='0';cEthNum!='5';cEthNum++)
	{
		snprintf(acFileName,sizeof(acFileName),"/sys/class/net/eth%c/address",cEthNum);
		if (NULL==(tEthFile=fopen(acFileName,"r")))
			continue;
		if (fgets(acMACAddress,sizeof(acMACAddress), tEthFile))
		{
			fclose(tEthFile);
			if (0 == memcmp(acMACAddress,WATCHGUARD_OUI,sizeof(WATCHGUARD_OUI)-1))
				//Found one, but no guarantee it is a Firebox II or III 
				break;
		}
	}
	//if ('5' == cEthNum)
		//Did not find a Watchguard OUI
		//return -1;
	//In order to make direct use of I/O ports from user space, access has to be requested
	// Tutorial: http://www.faqs.org/docs/Linux-mini/IO-Port-Programming.html
	// This needs "root" permission to succeed
	iRetVal = IOPERM(LED_BASEPORT, 3, 255);
	if (iRetVal) return iRetVal;

	//Reset all LEDs
	DrvSetLeds(0x0900);
	DrvSetLeds(0x0B00);
	DrvSetLeds(0x0300);
	DrvSetLeds(0x0700);
	DrvSetLeds(0x0F00);

	if (DRV_SLOW == uMode)
	{
		//Init time animation
		//One LED at a time for Status LEDs
		DrvSetLedsWait(LED_DISARMED,DRV_INIT_WAIT);
		DrvSetLedsWait(LED_ARMED,DRV_INIT_WAIT);
		DrvSetLedsWait(LED_SYS_A,DRV_INIT_WAIT);
		DrvSetLedsWait(LED_SYS_B,DRV_INIT_WAIT);
		DrvSetLeds(0x0900);
		//One LED at a time for Load LEDs
		for (u=1; u<256;u*=2)
			DrvSetLedsWait ((LED_LOAD_LO-1)|u,DRV_INIT_WAIT);
		DrvSetLeds(0x0B00);
		//One LED at a time for Traffic LEDs
		for (u=1; u<256;u*=2)
			DrvSetLedsWait ((LED_TRAFFIC_LO-1)|u,DRV_INIT_WAIT);
		DrvSetLeds(0x0300);
		//Triangle
		DrvSetLedsWait(LED_TRUST | LED_EXTRN | LED_OPTNL,DRV_INIT_WAIT);
		DrvSetLeds(LED_T2E_1 | LED_O2E_1);
		DrvSetLedsWait(LED_E2T_1 | LED_O2T_1| LED_T2O_1 | LED_E2O_1,DRV_INIT_WAIT);
		//Triangle
		DrvSetLeds(LED_T2E_2 | LED_O2E_2);
		DrvSetLedsWait(LED_E2T_2 | LED_O2T_2| LED_T2O_2 | LED_E2O_2,DRV_INIT_WAIT);
		DrvSetLeds(0x0700);
		DrvSetLeds(0x0F00);
	}
	return 0;
}

//DRIVER: Finalize function
//
static int DrvEnd(unsigned uCombo)
{
	//Reset all LEDs
	DrvSetLeds(0x0900);
	DrvSetLeds(0x0B00);
	DrvSetLeds(0x0300);
	DrvSetLeds(0x0700);
	DrvSetLeds(0x0F00);
	//Light left on, if needed
	if (uCombo)
		DrvSetLeds(uCombo);

	//This may need "root" access, and is done by Linux anyway on exit
	// may want to consider dropping this call if we change identity
	return IOPERM(LED_BASEPORT, 3, 0);
}

//DRIVER: Actually set the LEDs
//  No buffer, no validation
//
static inline void DrvSetLeds(unsigned uCombo)
{
	//uCombo is Group# and bits in 2 bytes
	OUTB((uCombo&0x00FF)^0xFF, LED_DATA);	//Put data for the LEDs, after inversion (bit at 1 means LED is "off")
	OUTB((uCombo>>8)|0x01, LED_CONTROL);	//Put Group # in control, Strobe bit to 1
	OUTB((uCombo>>8)&0xFE, LED_CONTROL);	//Flip Strobe bit to 0, causes Firebox II to output LED_DATA to LEDs
	OUTB((uCombo>>8)|0x01, LED_CONTROL);	//Flip Strobe bit to 1, causes Firebox III to output LED_DATA to LEDs
}

//DRIVER: Set LEDs and wait for a while
//
static void DrvSetLedsWait(unsigned uCombo, unsigned long ulInterval)
{
	const struct timespec tReq = { .tv_sec=0, .tv_nsec=ulInterval};
	struct timespec tRem = {.tv_sec=0, .tv_nsec=0};
	
	DrvSetLeds(uCombo);
	//Now wait for a while
	nanosleep(&tReq,&tRem);
	//Technically, can have time remaining in tRem
}

//CLIENT: This worker LED function updates the load LEDs based on system load average
//
static void DoLoad(void)
{
	static unsigned uLedBitsOld=0;
	char					acLoadAvg[256];
	unsigned			uLoadLastMinInt;
	unsigned			uLoadLastMinDec;
	unsigned 			uLoad;
	unsigned 			uLedBits;
	
	//Only get 1st number, load avg in last min
	//getloadavg(&dLoadLastMin, 1); //uClibc does not have this
	if (0 == GetInFile ("/proc/loadavg",acLoadAvg,sizeof(acLoadAvg)))
		return;
	//Try and avoid the use of floating point
	if (2 != sscanf(acLoadAvg,"%u.%u ",&uLoadLastMinInt, &uLoadLastMinDec))
		return;
	//Normalize, mapping load to number of LEDs
	uLoad=(unsigned)((uLoadLastMinInt*100+uLoadLastMinDec)/15);

	uLoad=(uLoad >= 128) ? 256 : uLoad<<1; //Anything too high will light all LEDs

	for (uLedBits=1;uLoad>1;uLedBits=uLedBits<<1) //Cheap log function
		uLoad=uLoad>>1;

	//Apply the style modifyer
	if (uStackStyle & STACK_LINE)
		uLedBits>>=1;
	if (uStackStyle & STACK_BAR || uStackStyle & STACK_RAW)
		uLedBits--;
	if (uStackStyle & STACK_REVERSE)
		uLedBits^=0xFF;
	
	//Only update LEDs if necessary
	if (uLedBits != uLedBitsOld)
	{
		uLedBitsOld=uLedBits;
		DrvSetLeds(LED_LOAD_LO-1+uLedBits);
	}
}

//CLIENT: This worker LED function updates the Traffic LEDs based on /proc/net/dev
//
static void DoTraffic(void)
{
	static unsigned 		uLedBitsOld=0;
	static long long		lAllPacketsPrior=0;
	static struct timeval	tWhenPrior = { .tv_sec=0, .tv_usec=0};
	
	FILE 								*tProcFile=NULL;
	char							acLineBuffer[2048]="";

	long long 					lRxPackets;
	long long 					lTxPackets;
	long long					lAllPackets;
	long							lRate;

	struct timeval				tWhenNow;

	unsigned 					uLedBits;

	if((tProcFile=fopen( "/proc/net/dev", "r" )) == NULL)
		return;
	//Read clock, no timezone
	if (gettimeofday(&tWhenNow,NULL))
	{
		fclose(tProcFile);
		return;
	}
	fgets(acLineBuffer, sizeof(acLineBuffer), tProcFile); //Throw away line 1
	fgets(acLineBuffer, sizeof(acLineBuffer), tProcFile); //Throw away line 2
	//Read all lines, 1 by 1
	for (lAllPackets=0;fgets(acLineBuffer, sizeof(acLineBuffer), tProcFile);)
         if(    strstr(acLineBuffer, "eth0:") != NULL
			|| strstr(acLineBuffer, "eth1:") != NULL
			|| strstr(acLineBuffer, "eth2:") != NULL
			|| strstr(acLineBuffer, "eth3:") != NULL
			|| strstr(acLineBuffer, "eth4:") != NULL)
	    {
			//Only parse lines for base hardware interfaces, ignore others like lo, bond, tun, VLAN ethx., pppoe
			//This is done so we avoid counting same packet multiple times
			//Could be boot time rename of interface eth0-> eth1, could be PCI card with eth4
            //Spacing is dynamic, sometimes interface name plus ":" is followed by a space, sometimes not
            if(*( strchr(acLineBuffer,':')+1) == ' ' )
                sscanf(acLineBuffer, "%*s %*u %Lu %*u %*u %*u %*u %*u %*u %*u %Lu %*u %*u %*u %*u %*u %*u",
                                  &lRxPackets, &lTxPackets ); //This is for lines like eth1:<space>123
            else
                sscanf(acLineBuffer, "%*s %Lu %*u %*u %*u %*u %*u %*u %*u %Lu %*u %*u %*u %*u %*u %*u",
                                  &lRxPackets, &lTxPackets ); //This is for lines like eth1:123
			lAllPackets+=lRxPackets+lTxPackets;
		}
	fclose(tProcFile);
	//If first time in, we have neither the old time stamp nor the old packet count
	if (0L != lAllPacketsPrior)
	{
		//Rate is number of packets since last time divided by time diff since last time
		//Normalize rate by dividing by 64
		// 64     packets per second is 1 LED
		lRate=(lAllPackets-lAllPacketsPrior)*1000000/64/((tWhenNow.tv_sec-tWhenPrior.tv_sec)*1000000+tWhenNow.tv_usec-tWhenPrior.tv_usec);
		// 8192 packets per second and above is 8 LEDs
		if (lRate > 8192/64)
			lRate=8192/64;  //Anything higher than 8192 packets/second is all LEDs
		//compute the logarithm of the rate the cheap way
		//basically, find the highest bit set to 1
		for (uLedBits=1;lRate>0;uLedBits=uLedBits<<1)
			lRate=lRate>>1;
		//Apply the style modifyer
		if (uStackStyle & STACK_LINE)
			uLedBits>>=1;
		if (uStackStyle & STACK_BAR)
			uLedBits--;
		if (uStackStyle & STACK_RAW)
			uLedBits=(lAllPackets/64)&0xFF; //This displays the low byte of 64 packet _count_ (not _rate_) in binary
		if (uStackStyle & STACK_REVERSE)
			uLedBits^=0xFF;
		//Only update LEDs if necessary
		if (uLedBits != uLedBitsOld)
		{
			uLedBitsOld=uLedBits;
			//Use thread-safe LED update
			DrvSetLeds(LED_TRAFFIC_LO-1+uLedBits);
		}
	}
	//Set time and count baseline for next run
	lAllPacketsPrior=lAllPackets;
	tWhenPrior=tWhenNow;
}

//CLIENT: This worker LED function blinks a LED, heartbeat sort of thing
//
static void DoBlink(void)
{
	static unsigned uStatusLEDs=LED_ARMED | LED_SYS_A;

	//Flip LED_ARMED
	uStatusLEDs^=LED_ARMED&0x00FF;
	DrvSetLeds(uStatusLEDs);
}
//CLIENT: Helper function to get stuff in text files
//
static int GetInFile(const char *acFileName, char *acDest, const unsigned uSize)
{
	FILE		*tFile=NULL;
	int		iReturnVal=-1;

	if (NULL==(tFile=fopen(acFileName,"r")))
		return 0;
	if (NULL == fgets(acDest, uSize, tFile))
		iReturnVal=0;
	fclose(tFile);
	return iReturnVal;
}

//CLIENT: scheduler
// This function does the infinite loop of wait-run-wait for all workers
//
void Scheduler(tExecTable ptWorkTable[], unsigned uCount)
{
	struct timespec tReq = { .tv_sec=0, .tv_nsec=WRK_WAIT};
	struct timespec tRem = {.tv_sec=0, .tv_nsec=0};
	unsigned	u;

	//Run init code, if present
	for (u=0;u<uCount;u++)
		 if (ptWorkTable[u].pfConstr)
			 (ptWorkTable[u].pfConstr)();

	//loop until global variable changes to 0
	for (;uKeepGoing;)
	{
		//Workers
		for (u=0;u<uCount;u++)
			if (0 == ptWorkTable[u].uRunningCount)
			{
				ptWorkTable[u].uRunningCount=ptWorkTable[u].uSkipCount;
				(*ptWorkTable[u].pfCode)();
			}
			else
				ptWorkTable[u].uRunningCount--;
		//Wait a while
		nanosleep(&tReq,&tRem);
	}

	//Cleanup code, if present
	for (u=0;u<uCount;u++)
		 if (ptWorkTable[u].pfDestr)
			 (*ptWorkTable[u].pfDestr)();
}

//CLIENT: This is the handler for all termination signals.
// Not much can be done here, so just flip a global variable
// to signal worker threads to stop
//
void ExitHandler(int iSigNum)
{
	//Just change flag for any termination signal
	uKeepGoing=0;
}

//CLIENT: This is the handler for the USR signals
// A good place to handle alternative styles
//
void UserHandler(int iSigNum)
{
	//For USR1 and USR2, flip styles
	switch (iSigNum)
	{
		case SIGUSR1:
			if (uStackStyle & STACK_BAR)
			{
				uStackStyle^=STACK_BAR;
				uStackStyle|=STACK_LINE;
				return;
			}
			if (uStackStyle & STACK_LINE)
			{
				uStackStyle^=STACK_LINE;
				uStackStyle|=STACK_RAW;
				return;
			}
			if (uStackStyle & STACK_RAW)
			{
				uStackStyle^=STACK_RAW;
				uStackStyle|=STACK_BAR;
				return;
			}
			//break;
			
		case SIGUSR2:
			uStackStyle^=STACK_REVERSE;
	}
}

//CLIENT: Entry point upon execution
//
int main(int nArgc, char **asArgv)
{
	puts(PACKAGE_STRING);

	//Initialize Driver, requires "root"
	if (DrvInit(DRV_SLOW))
	{
		puts("Cannot Initialize Driver...");
		return errno;
	}
	//Setup handler for all termination signals (Control-c, Control-Backslash, or kill)
	signal(SIGINT, ExitHandler);
	signal(SIGTERM, ExitHandler);
	signal(SIGQUIT, ExitHandler);
	//Setup handler for USR1, USR2 "kill -USR1 <pid>, kill -USR2 <pid>"
	signal(SIGUSR1, UserHandler);
	signal(SIGUSR2, UserHandler);
	//Run the worker scheduler
	Scheduler(tLEDProcTable,SIZE(tLEDProcTable));

	//Finalize, and turn the DISARMED light on
	if (DrvEnd(LED_DISARMED))
	{
		puts("Cannot Finalize Driver...");
		return errno;
	}
	return 0;
}


fbled.h
//           // fbled.h
//
// Copyright (C) 2011 - F Mertz
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

//  Wed March 16
//  Copyright  2011  F Mertz
//  <fireboxled@gmail.com>
//
// This project aims at providing a simple deamon to update the front LEDs on a Watchguard Firebox II/III. This code is
// logically separated into a Driver and a Client.
//
// The Driver is responsible for updating the LEDs based on input parameters, and deals with low-level I/O ports.
// This driver might be reimplemented as a real Linux driver/module at a later point.
//
// The Client is responsible for gathering the live values from the running system, parse them, normalize them, then pass them
// to the Driver.
//=============================================================================
//
// Driver Logic:
//  The LEDs are controlled similarly to controlling a printer on the good old PC parallel port.
//  The general idea is that there are a number of on/off lights that need to be controlled individually. This is done by sending
//  data to specific "ports", 8-bit at a time. As we have more than 8 LEDs, there is the issue of addressing the LEDs and the
//  issue of controlling them. This mapping has been reverse-engineered by Jani Laaksonen.
//  The short version is that the data port, BASEPORT+0, controls a group of up to 8 LEDs, one LED for each bit.
//  What group of 8 LEDs the data is for is controlled by the control port, BASEPORT+2. The actual updating of the LEDs is
//  a 3 steps process:
//  1.The control port is updated with the LED Group Number
//  2.The LED data is placed in the data port, 1 bit for each LED
//  3.Control bit-0 of the control port is then flipped to actually change the LED
//
//  The mapping of LED Group to Control bits is:
//		Group 0, 0x02->0x03, This is for the 4 Status LEDs
//		Group 1, 0x00->0x01, This is for 8 Load LEDs
//		Group 2, 0x08->0x09, This is for 8 Traffic LEDs
//		Group 3, 0x0c->0x0d, This is for part of the triangle (7 LEDs)
//		Group 4, 0x04->0x05, This is for the rest of the triangle (8 LEDs)
//
//  Last, there is hardware inversion for bits 1011, so these bits need to be flipped.
//                      See http://en.wikipedia.org/wiki/Parallel_port
//
//  Note: LEDs are controlled individually, there is no grouping of any kind, i.e. the 8 LEDs for Traffic or Load can
//            be lit up individually, there is no "stack" concept in hardware. It all need to be built in the code.
//  Note: As is tradition, the Driver is just responsible for offering the means to control the LEDs. It is not responsible for
//            policy, i.e. what information in what form goes to what LED.
//  Note: This is a Linux implementation.
//=============================================================================
//
// Client Logic:
//  The LEDs are labelled in a pretty generic way, and can be used to convey a number of statuses. This Client imlements
//  the policies described below. The original policies from the manufacturer are quoted.

//  Disarm: "Red light indicates the Firebox detected an error, shut down its interfaces, and will not forward any packets. Reboot the Firebox."
//               Idea: ready to poweroff, root file system unmounted
//  Armed: "Green light indicates the Firebox has been booted and is running."
//               Idea: WAN has IP address.
//               Idea: Firewall has at least 1 entry
//               Idea: the boot process has completed
//               Idea: flashes to indictate this code is running
//    Sys A: "Indicates that the Firebox is running from its primary user-defined configuration."
//               Idea: os running off of ramdrive, early boot
//    Sys B: "Indicates that the Firebox is running from the readonly factory default system area."
//               Idea: os running off of storage device, root file system mounted, normal operation
//   Power: "Indicates that the Firebox is currently powered up."
//               Direct harware wiring, no software control
// Triangle: "Indicates traffic between Firebox interfaces. Green arrows briefly light to indicate allowed traffic
//                between two interfaces in the direction of the arrows. A red light at a triangle corner indicates 
//                that the Firebox is denying packets at that interface."
//                For External, Trusted and Other, off if no link, on if link AND IP address, blink if link, no IP
//    Traffic: "A stack of lights that functions as a meter to indicate levels of traffic volume through the 
//                Firebox. Low volume indicators are green, while high volume indicators are yellow. The display 
//                updates three times per second. The scale is exponential: the first light represents 64 packets/ 
//                second, the second light represents 128 packets/second, increasing to the eighth light which 
//                represents 8,192 packets/second. "
//                This stack is controlled by /proc/net/dev, with a logarithmic scale.
//     Load: "A stack of lights that functions as a meter to indicate the system load average. The system load 
//               average is the average number of processes running (not including those in wait states) during 
//               the last minute. Low average indicators are green, while high average indicators are yellow. The 
//               display updates three times per second. The scale is exponential with each successive light representing 
//               a doubling of the load average. The first light represents a load average of 0.15. The most 
//               significant load factor on a Firebox is the number of proxies running."
//               This stack is controlled by /proc/loadavg, with a logarithmic scale. The first number is the number of processes
//               actually using CPU cycles to run during the last minute.
//
// Client Architecture
//               Overall, the client is basically an infinite loop of execution threads with a configurable wait time. The
//               implementation is based on multi-threading a number of functions (the Do... functions), one per group
//               of LEDs. The exit condition is the receipt of one of the usual termination signals.
//=============================================================================

//Driver Constants
#define LED_BASEPORT	0x378
#define LED_DATA			(LED_BASEPORT+0)
#define LED_CONTROL		(LED_BASEPORT+2)
#define LED_WAITPORT		0x80 //This is used when using the "slow" outb_p

//LED Coding; Control bits in 1 byte followed by Data bits for each LED
//LED Group 0: Status LEDs, Control Code 2, 0010 xor'ed with 1011 is 1001, 0x09
#define LED_DISARMED	0x0901
#define LED_ARMED			0x0902
#define LED_SYS_A 			0x0904
#define LED_SYS_B			0x0908
#define LED_ENABLE		0x0940

//LED Group 1: Load, Control Code 0, 0000 xor'ed with 1011 is 1011, 0x0B
#define LED_LOAD_LO		0x0B01
#define LED_LOAD_HI		0x0BFF

//LED Group 2: Traffic, Control Code 8, 1000 xor'ed with 1011 is 0011, 0x03
#define LED_TRAFFIC_LO	0x0301
#define LED_TRAFFIC_HI	0x03FF
     
//LED Group 3: Triangle, Control Code C, 1100 xor'ed with 1011 is 0111, 0x07
#define LED_T2E_1			0x0701
#define LED_T2E_2			0x0702
#define LED_O2E_1			0x0704
#define LED_O2E_2			0x0708
#define LED_EXTRN			0x0710
#define LED_TRUST			0x0720
#define LED_OPTNL			0x0740

//LED Group 4: Triangle, Control Code 4, 0100 xor'ed with 1011 is 1111, 0x0F
#define LED_E2T_1			0x0F01
#define LED_E2T_2			0x0F02
#define LED_O2T_1			0x0F04
#define LED_O2T_2			0x0F08
#define LED_T2O_1			0x0F10
#define LED_T2O_2			0x0F20
#define LED_E2O_1			0x0F40
#define LED_E2O_2			0x0F80

//Driver constants
#define DRV_FAST			0
#define DRV_SLOW			1
#define DRV_INIT_WAIT		200000000 //.2 sec

//Workers constants
#define WRK_WAIT			1000000000L/6	//Base wait time is 1/6 of a sec

#ifndef LED_DEBUG
//Real Watchguard range
#define WATCHGUARD_OUI "00:90:7f"
#else
//Virtualbox range
#define WATCHGUARD_OUI "08:00:27"
#endif

//Stack styles for Traffic and load modifyer
#define STACK_BAR			1
#define STACK_LINE			2
#define STACK_RAW			4
#define STACK_REVERSE	8

//Driver prototypes
static int DrvInit(unsigned char);
static int DrvEnd(unsigned);
static inline void DrvSetLeds(unsigned);
static void DrvSetLedsWait(unsigned, unsigned long);

//Client Structures
struct ExecTable
{
	void				(*pfConstr)(void);						//Contructor, init code, if needed
	void 			(*pfCode)(void);					    	//Code to run
	void				(*pfDestr)(void);							//Destructor, close code, if needed
	unsigned	uSkipCount;
	unsigned	uRunningCount;
};

typedef struct ExecTable tExecTable;

//Client Contants

//Client Prototypes
static void DoLoad(void);
static void DoTraffic(void);
static void DoBlink(void);
static int GetInFile(const char *, char *, const unsigned );
void Scheduler(tExecTable *, unsigned);
void ExitHandler(int);
void UserHandler(int);

//Debugging Assist
#ifndef LED_DEBUG
//Run-time version of low-level io routines
#define OUTB(port,val) outb((port),(val))
#define IOPERM(port, count, yesno) ioperm((port), (count), (yesno))
#else
//Debug version of low-level routines
#define OUTB(val,port) printf("outb(%04X,%02X)\n",port,val)
#define IOPERM(port, count, yesno) (printf("ioperm(%04X,%u)\n",port,count)==0)
#endif

#define SIZE(a) sizeof(a)/sizeof(a[0])
